File: 20250502/파괴되지_않은_건물.java
import java.util.*;

class Solution {
    public int solution(int[][] board, int[][] skills) {
        int[][] arr = new int[board.length + 1][board[0].length + 1];

        for (int[] skill : skills) {
            int type = skill[0];
            int r1 = skill[1], c1 = skill[2];
            int r2 = skill[3], c2 = skill[4];
            int degree = skill[5];

            int amount = type == 1 ? -(degree) : degree;

            arr[r1][c1] += amount;
            arr[r1][c2 + 1] += -(amount);
            arr[r2 + 1][c1] += -(amount);
            arr[r2 + 1][c2 + 1] += amount;
        }

        // 누적합 계산
        for (int r = 0; r < board.length; r++) {
            for (int c = 0; c < board[0].length; c++) arr[r][c + 1] += arr[r][c];
        }

        for (int c = 0; c < board[0].length; c++) {
            for (int r = 0; r < board.length; r++) arr[r + 1][c] += arr[r][c];
        }

        // 스킬 사용 내역 반영
        for (int r = 0; r < board.length; r++) {
            for (int c = 0; c < board[0].length; c++) board[r][c] += arr[r][c];
        }

        return Arrays.stream(board)
                .mapToInt(b -> Arrays.stream(b).map(i -> i > 0 ? 1 : 0).sum())
                .sum();
    }
}

File: 20250502/파괴되지_않은_건물.js
function solution(board, skill) {
    const N = board.length;
    const M = board[0].length;

    const effect = Array.from({ length: N + 1 }, () => Array(M + 1).fill(0));

    // 누적합 배열에 차등 적용
    for (let [type, r1, c1, r2, c2, degree] of skill) {
        const delta = type === 1 ? -degree : degree;

        effect[r1][c1] += delta;
        if (c2 + 1 < M) effect[r1][c2 + 1] -= delta;
        if (r2 + 1 < N) effect[r2 + 1][c1] -= delta;
        if (r2 + 1 < N && c2 + 1 < M) effect[r2 + 1][c2 + 1] += delta;
    }

    // 가로 누적합
    for (let i = 0; i < N; i++) {
        for (let j = 1; j < M; j++) {
            effect[i][j] += effect[i][j - 1];
        }
    }

    // 세로 누적합
    for (let j = 0; j < M; j++) {
        for (let i = 1; i < N; i++) {
            effect[i][j] += effect[i - 1][j];
        }
    }

    // 실제 적용해서 내구도 계산
    let count = 0;
    for (let i = 0; i < N; i++) {
        for (let j = 0; j < M; j++) {
            if (board[i][j] + effect[i][j] > 0) {
                count++;
            }
        }
    }

    return count;
}
